File: Leetcode/Algorithms/Medium/HandOfStraights.cpp
class Solution
{
public:
    bool isNStraightHand(vector<int> &hand, int W)
    {
        if (hand.size() % W != 0)
            return false;
        map<int, int> m;

        for (int curr : hand)
            m[curr]++;

        while (m.size() != 0)
        {
            int current = m.begin()->first;
            for (int i = 0; i < W; i++)
            {
                if (m[current + i] == 0)
                    return false;
                else if (--m[current + i] < 1)
                    m.erase(current + i);
            }
        }
        return true;
    }
};

// Time Complexity - O(N * W)
// Space Complexity - O(N), N is the size of hand

File: Leetcode/Algorithms/Medium/HandOfStraights/HandOfStraights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        if(hand.length % groupSize != 0) return false; 
        
        TreeMap<Integer, Integer> card_counts = new TreeMap<Integer, Integer>(); 
        
        for(int card: hand) 
        {
            if(!card_counts.containsKey(card))
            {
                card_counts.put(card, 1); 
            }
            else 
            {
                card_counts.replace(card, card_counts.get(card) + 1);
            }
        }
        
        while(card_counts.size() > 0)
        {
            int min_val = card_counts.firstKey(); 
            for(int card = min_val; card < min_val + groupSize; card++)
            {
                if(!card_counts.containsKey(card)) return false;
                int count = card_counts.get(card); 
                if(count == 1) card_counts.remove(card); 
                else 
                card_counts.replace(card, count - 1); 
            }
            
        }
        return true; 
    }
}

// Time Complexity - O(N * W), N is the size of hand and W is the groupSize 
// Space Complexity - N, N is the size of hand